### PR TITLE
Mapgen: Optimise cave noises and tunnel excavation

### PR DESCRIPTION
**Pull request by paramat**
Instead of doing nothing at node_max.Y + 1 use 1-down
overgeneration for tunnel generation and noisemaps
Remove some old unused code from mgv7
//////////////////////////////////

In #3850 i made tunnel generation do nothing in the node_max.Y + 1 layer of nodes, while still looping through that y-vlaue and still calculating noise for it.
Here i make the obvious optimisation of not doing these unnecessary actions, cave noise perlinmaps are calculated for 1-down overgeneration, saving large amounts of calculation.

Changes to mgv5 come later as that's a more complex change.

### CODE REVIEW

**Comment by paramat** [comment]
Tested.

**Comment by kwolekr** [comment]
You can't approve your own PR.

**Comment by C1ffisme** [comment]
@kwolekr Actually, right now, you can.

I highly disapprove of it, but you can do it.

**Comment by celeron55** [comment]
You can approve your own PR if you want to; there's intentionally no rule that says otherwise. If it weren't that way, then a core dev's PR essentially would always require three approvals, which I think is excessive.

**Comment by paramat** [comment]
We've been allowing core-dev self-approval for months. I think little would get done otherwise.

**Comment by sfan5** [comment]
:+1:

**Comment by kwolekr** [comment]
Where are people getting these 'rules' from?  Is it written somewhere?
Common sense says you can't approve your own code, the point of a code review is for others to review your own code to pick out what was missed.

File: src/mapgen_v7.cpp [hunk 1]
```diff
@@ -451,41 +456,6 @@ bool MapgenV7::getMountainTerrainFromMap(int idx_xyz, int idx_xz, s16 y)
 }
 
 
-#if 0
-void MapgenV7::carveRivers() {
```
Comment (kwolekr) [inline src/mapgen_v7.cpp]: Why did you delete this?

**Comment by paramat** [comment]
"1. You can push something to upstream only if two members of the core team agree on it."
I don't think that means 'You' is person 1 and 'two members' are 2 other people. If it read 'agree to it' that might imply two other people.
I'm okay with disallowing self-approval as long as then the number of approvers required is reduced to 1, these are just 2 ways of stating the same thing: 2 core devs agree on a PR, one of those may be the author.
If the author's opinion of their own PR doesn't count towards the '2' required then that's weird as the author is often the person most knowledgeable about what's going on, for example my own PRs on mgfractal.

**Comment by paramat** [comment]
kwolekr i need to ask you about the removal of old unused code, do you want it kept? I've kept it in there for a long time but it is a little irritating because there's so much of it, it makes the document longer and means more scrolling around.
Mgv7 has settled and carveRivers will not be used here, addTopNodes seems to be for use with the previous mgv7 version, i can't see any use for either in this file, if you want them kept i would like to move them to the end of the file so i don't have to scroll past them.

**Comment by kwolekr** [comment]
The approval discussion requires a lot of discussion and probably best done on IRC.

As for deleting removed code... I'd rather not, if it gets in the way, does it help to move it to the bottom of the source file?

**Comment by paramat** [comment]
No problem, i'll move it to file end.

**Comment by paramat** [comment]
Done.

File: src/mapgen_flat.cpp [hunk 2]
```diff
@@ -58,7 +58,10 @@ MapgenFlat::MapgenFlat(int mapgenid, MapgenParams *params, EmergeManager *emerge
 	//// amount of elements to skip for the next index
 	//// for noise/height/biome maps (not vmanip)
 	this->ystride = csize.X;
-	this->zstride = csize.X * (csize.Y + 2);
+	// 1-down overgeneration
+	this->zstride_1d = csize.X * (csize.Y + 1);
+	// 1-up 1-down overgeneration
+	this->zstride_1u1d = csize.X * (csize.Y + 2);
```
Comment (kwolekr) [inline src/mapgen_flat.cpp]: Where is this variable used?

File: src/mapgen_flat.cpp [hunk 3]
```diff
@@ -566,19 +570,18 @@ void MapgenFlat::generateCaves(s16 max_stone_y)
...
 		// Biome of column
 		Biome *biome = (Biome *)bmgr->getRaw(biomemap[index2d]);
 
-		for (s16 y = node_max.Y + 1; y >= node_min.Y - 1;
+		// Don't excavate the overgenerated stone at node_max.Y + 1,
+		// this creates a 'roof' over the tunnel, preventing light in
+		// tunnels at mapchunk borders when generating mapchunks upwards.
+		// This 'roof' is removed when the mapchunk above is generated.
+		for (s16 y = node_max.Y; y >= node_min.Y - 1;
 				y--, index3d -= ystride, vm->m_area.add_y(em, vi, -1)) {
```
Comment (kwolekr) [inline src/mapgen_flat.cpp]: Since you're not using a continue statement in this loop anymore, could you put the other variable modifications on separate lines at the end of the loop?  Much less confusing and looks cleaner IMO.

File: src/mapgen_flat.cpp (hunk 2 above)
Comment (paramat) [inline src/mapgen_flat.cpp]: In some mapgens only one type of zstride is used, but i kept both as i thought that good practice because until recently zstride was present but not used in mgv7.
However agreed i will remove the unused variables.

File: src/mapgen_flat.cpp (hunk 3 above)
Comment (paramat) [inline src/mapgen_flat.cpp]: I recently caused a bug in mgwatershed by adding a continue when the modifications were at loop end, so i consider it good practice to keep them at the start, however i will put them on seperate lines to be clearer.

Comment (paramat) [inline src/mapgen_flat.cpp]: Actually continue statements are still used in this loop.

**Comment by paramat** [comment]
Updated with line comments attended to.

**Comment by paramat** [comment]
Review and approval from kwolekr http://irc.minetest.ru/minetest-dev/2016-04-08#i_4578785

**Comment by paramat** [comment]
465bb6f5d13803819f329e4920bceb25a3887c6f
